### DISCUSSION TITLE
lazy_cache功能和smartdns的serve-expired功能类似吗

### DISCUSSION THREAD

**Discussion by yeezylife** [IrineSistiana/mosdns#172] [Q&A] [answered]
如果是这样的话,那按照smartdns作者推荐的serve-expired-ttl=1~2天的值来设置lazy_cache_ttl是合理的吗?

补一下serve-expired功能的描述:

"DNS结果缓存有过期时间，这个过期时间就是dns结果的ttl值，默认情况下，当ttl小于0时，缓存就失效了，有新查询时，就要重新到远端服务器再次查询结果，这段时间就要客户端等待。

serve-expired的功能就是避免这个查询等待：原理是大部分情况，域名对应的ip地址变化是很低概率的，用过期的缓存也是没问题的，所以优化策略就是当缓存ttl小于0时，缓存不失效，变为过期缓存，当有客户机查询dns时，返回这个过期的缓存，并将结果的ttl设置为0，要求客户机下次再次查询，同时smartdns后台发送请求给上游更新缓存。"

**Comment by IrineSistiana** [comment, marked as answer]
一模一样。

**Reply by yeezylife** [reply]
> 一模一样。

延迟高的服务器,现在变秒开了,这功能挺好.
是不是可以适当扩展一下,我看smartdns和unbound围绕这个功能还能提供以下设置:
serve-expired-ttl-reset=yse/no
serve-expired-client-timeout=
serve-expired-reply-ttl=

**Reply by IrineSistiana** [reply]
serve-expired-reply-ttl 加了。其他选项感觉不常用。

~懒~

**Reply by yeezylife** [reply]
就是提供一点思路,不常用的不加也没事,加油!

**Comment by yeezylife** [comment]
开启了lazy_cache功能之后是不是没必要再用ttl插件改ttl了啊
我lazy_cache_ttl的值为两天,而ttl插件改最小TTL值为1小时
这样我的ttl插件改的最小值是不是等于白用了(1小时<两天)

**Comment by yeezylife** [comment]
好像还是有不同,改了ttl之后ANSWERS:还是有变化
    QUESTIONS:
        baidu.com, type = A, class = IN
    ANSWERS:
    ->  baidu.com
        internet address = 220.181.38.148
        ttl = 3600 (1 hour)
    ->  baidu.com
        internet address = 220.181.38.251
        ttl = 3600 (1 hour)
    AUTHORITY RECORDS:
    ->  baidu.com
        nameserver = ns7.baidu.com
        ttl = 86377 (23 hours 59 mins 37 secs)
    ->  baidu.com
        nameserver = ns4.baidu.com
        ttl = 86377 (23 hours 59 mins 37 secs)
    ->  baidu.com
        nameserver = dns.baidu.com
        ttl = 86377 (23 hours 59 mins 37 secs)
    ->  baidu.com
        nameserver = ns3.baidu.com
        ttl = 86377 (23 hours 59 mins 37 secs)
    ->  baidu.com
        nameserver = ns2.baidu.com
        ttl = 86377 (23 hours 59 mins 37 secs)
    ADDITIONAL RECORDS:
    ->  ns7.baidu.com
        internet address = 180.76.76.92
        ttl = 171948 (1 day 23 hours 45 mins 48 secs)
    ->  ns4.baidu.com
        internet address = 14.215.178.80
        ttl = 171946 (1 day 23 hours 45 mins 46 secs)
    ->  dns.baidu.com
        internet address = 110.242.68.134
        ttl = 171942 (1 day 23 hours 45 mins 42 secs)
    ->  ns2.baidu.com
        internet address = 220.181.33.31
        ttl = 85698 (23 hours 48 mins 18 secs)
    ->  ns3.baidu.com
        internet address = 112.80.248.64
        ttl = 85672 (23 hours 47 mins 52 secs)
